File: wxyc-etl-python/python/wxyc_etl/logger.py
"""Sentry + structured JSON logging for WXYC ETL pipelines (Python side).

Mirrors :mod:`wxyc_etl::logger` (Rust). Every Python ETL should call
:func:`init_logger` once at startup with a ``repo`` and ``tool`` identifying
the pipeline. Logs emit one JSON object per line on the root logger with the
tags ``repo``, ``tool``, ``step``, and ``run_id`` (``step`` is set per-call via
:func:`logging.Logger.info` ``extra=`` or via :class:`LoggerStep`). When
``SENTRY_DSN`` is set in the environment (or supplied to :func:`init_logger`)
unhandled exceptions and ``logger.error`` events are forwarded to Sentry with
the same tags.

Usage::

    from wxyc_etl.logger import init_logger

    guard = init_logger(repo="discogs-etl", tool="discogs-etl daily-sync")

    import logging
    log = logging.getLogger(__name__)
    log.info("loaded recordings", extra={"step": "import", "rows": 42})

The returned :class:`LoggerGuard` flushes pending Sentry events when its
``flush`` method is called (or when the program exits).
"""

from __future__ import annotations

import atexit
import logging
import os
import sys
import uuid
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class LoggerConfig:
    """Configuration for :func:`init_logger`. ``sentry_dsn=None`` falls back to
    the ``SENTRY_DSN`` env var; ``run_id=None`` generates a UUIDv4."""

    repo: str
    tool: str
    sentry_dsn: Optional[str] = None
    run_id: Optional[str] = None


class LoggerGuard:
    """Returned by :func:`init_logger`. Call :meth:`flush` (or rely on the
    ``atexit`` hook) to drain Sentry before the process exits.
    """

    def __init__(self, run_id: str, sentry_enabled: bool) -> None:
        self.run_id = run_id
        self.sentry_enabled = sentry_enabled

    def flush(self, timeout: float = 2.0) -> None:
        if not self.sentry_enabled:
            return
        try:
            import sentry_sdk

            client = sentry_sdk.Hub.current.client
            if client is not None:
                client.flush(timeout=timeout)
        except Exception:  # pragma: no cover - flush is best-effort
            pass


class _ContextFilter(logging.Filter):
    """Inject repo/tool/run_id onto every record so the JSON formatter emits
    them. Per-event ``step`` is supplied by callers via ``extra={"step": ...}``
    and survives this filter unchanged.
    """

    def __init__(self, repo: str, tool: str, run_id: str) -> None:
        super().__init__()
        self._repo = repo
        self._tool = tool
        self._run_id = run_id

    def filter(self, record: logging.LogRecord) -> bool:
        if not hasattr(record, "repo"):
            record.repo = self._repo
        if not hasattr(record, "tool"):
            record.tool = self._tool
        if not hasattr(record, "run_id"):
            record.run_id = self._run_id
        if not hasattr(record, "step"):
            record.step = ""
        return True


_INITIALIZED = False


def init_logger(
    repo: str,
    tool: str,
    sentry_dsn: Optional[str] = None,
    run_id: Optional[str] = None,
    level: int = logging.INFO,
) -> LoggerGuard:
    """Initialize Sentry (when DSN is available) and JSON logging on the root
    logger. Idempotent: subsequent calls update tags but don't double-install
    handlers.
    """

    global _INITIALIZED

    rid = run_id or str(uuid.uuid4())
    dsn = sentry_dsn if sentry_dsn is not None else os.environ.get("SENTRY_DSN")
    sentry_enabled = bool(dsn)

    if sentry_enabled:
        import sentry_sdk

        sentry_sdk.init(
            dsn=dsn,
            traces_sample_rate=0.0,
            attach_stacktrace=True,
        )
        sentry_sdk.set_tag("repo", repo)
        sentry_sdk.set_tag("tool", tool)
        sentry_sdk.set_tag("run_id", rid)

    if not _INITIALIZED:
        try:
            from pythonjsonlogger.json import JsonFormatter
        except ImportError:
            try:
                from pythonjsonlogger.jsonlogger import (  # type: ignore[no-redef]
                    JsonFormatter,
                )
            except ImportError as exc:  # pragma: no cover - dep is declared
                raise RuntimeError(
                    "wxyc_etl.logger requires python-json-logger; "
                    "ensure the wxyc-etl wheel's runtime deps are installed"
                ) from exc

        formatter = JsonFormatter(
            "%(asctime)s %(levelname)s %(name)s %(message)s "
            "%(repo)s %(tool)s %(step)s %(run_id)s",
            rename_fields={"asctime": "timestamp", "levelname": "level"},
        )

        handler = logging.StreamHandler(sys.stderr)
        handler.setFormatter(formatter)
        handler.addFilter(_ContextFilter(repo, tool, rid))

        root = logging.getLogger()
        root.addHandler(handler)
        root.setLevel(level)

        guard = LoggerGuard(rid, sentry_enabled)
        atexit.register(guard.flush)
        _INITIALIZED = True
        return guard

    # Already initialized: reset the context filter on existing handlers so
    # the new repo/tool/run_id apply going forward.
    new_filter = _ContextFilter(repo, tool, rid)
    for handler in logging.getLogger().handlers:
        # Drop any prior _ContextFilter to avoid duplicate-tag injection.
        for existing in list(handler.filters):
            if isinstance(existing, _ContextFilter):
                handler.removeFilter(existing)
        handler.addFilter(new_filter)

    return LoggerGuard(rid, sentry_enabled)


__all__ = ["LoggerConfig", "LoggerGuard", "init_logger"]

File: wxyc-etl/src/logger.rs
//! Sentry + structured JSON logging for WXYC ETL pipelines.
//!
//! Every ETL binary should call [`init`] once at startup with a [`LoggerConfig`]
//! identifying the repo and tool. Logs emit as one JSON object per line with the
//! tags `repo`, `tool`, `step`, and `run_id` (the last set per-span via
//! `tracing` instrumentation). When `SENTRY_DSN` is set in the environment (or
//! provided in the config) panics and `tracing::error!` events are forwarded to
//! Sentry with the same tags.
//!
//! ## Usage
//!
//! ```no_run
//! use wxyc_etl::logger::{self, LoggerConfig};
//!
//! fn main() {
//!     let _guard = logger::init(LoggerConfig {
//!         repo: "musicbrainz-cache",
//!         tool: "musicbrainz-cache build",
//!         sentry_dsn: None,
//!         run_id: None,
//!     });
//!
//!     tracing::info_span!("import", step = "import").in_scope(|| {
//!         tracing::info!(rows = 42, "loaded recordings");
//!     });
//! }
//! ```

use std::sync::OnceLock;

use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;

/// Configuration for [`init`]. Pass `None` for `sentry_dsn` to fall back to the
/// `SENTRY_DSN` environment variable; pass `None` for `run_id` to generate a
/// fresh UUIDv4 for the run.
pub struct LoggerConfig {
    /// Repository name. Set as the Sentry tag `repo`.
    pub repo: &'static str,
    /// Tool name (typically `<repo> <subcommand>`). Set as the Sentry tag `tool`.
    pub tool: &'static str,
    /// Sentry DSN. `None` falls back to `SENTRY_DSN` env var; if neither is set
    /// Sentry is disabled but JSON logging still initializes.
    pub sentry_dsn: Option<String>,
    /// Run identifier. `None` generates a UUIDv4. Set as the Sentry tag `run_id`.
    pub run_id: Option<String>,
}

/// RAII guard returned by [`init`]. Drop it at the end of `main` to flush
/// pending Sentry events before exit.
pub struct LoggerGuard {
    _sentry: Option<sentry::ClientInitGuard>,
}

static INIT: OnceLock<()> = OnceLock::new();

/// Initialize Sentry and JSON logging. Safe to call once per process; later
/// calls are a no-op (and return a `LoggerGuard` whose Sentry slot is empty).
pub fn init(config: LoggerConfig) -> LoggerGuard {
    let run_id = config
        .run_id
        .unwrap_or_else(|| uuid::Uuid::new_v4().to_string());

    let dsn = config
        .sentry_dsn
        .or_else(|| std::env::var("SENTRY_DSN").ok());

    let mut sentry_guard = None;
    if let Some(dsn) = dsn.as_deref().filter(|s| !s.is_empty()) {
        let guard = sentry::init((
            dsn.to_string(),
            sentry::ClientOptions {
                release: sentry::release_name!(),
                attach_stacktrace: true,
                ..Default::default()
            },
        ));
        sentry::configure_scope(|scope| {
            scope.set_tag("repo", config.repo);
            scope.set_tag("tool", config.tool);
            scope.set_tag("run_id", &run_id);
        });
        sentry_guard = Some(guard);
    }

    if INIT.set(()).is_ok() {
        let _ = tracing_log::LogTracer::init();

        let env_filter = tracing_subscriber::EnvFilter::try_from_default_env()
            .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("info"));

        // Write to stderr so the contract matches `env_logger` and POSIX
        // convention: stdout is for the program's data output, stderr is for
        // diagnostics. Existing tests across consumer repos read stderr for
        // log lines and break if logs land on stdout instead.
        let json_layer = tracing_subscriber::fmt::layer()
            .json()
            .with_current_span(true)
            .with_span_list(false)
            .with_target(true)
            .with_writer(std::io::stderr);

        let registry = tracing_subscriber::registry()
            .with(env_filter)
            .with(json_layer);

        if sentry_guard.is_some() {
            let _ = registry.with(sentry_tracing::layer()).try_init();
        } else {
            let _ = registry.try_init();
        }

        // Re-set repo/tool/run_id as tracing fields on the root span so every
        // event carries them even outside a user-defined span.
        let span = tracing::info_span!(
            "wxyc_etl",
            repo = %config.repo,
            tool = %config.tool,
            run_id = %run_id,
        );
        let _ = span.entered();
    }

    LoggerGuard {
        _sentry: sentry_guard,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn init_without_dsn_does_not_panic() {
        // Safety net: explicitly clear SENTRY_DSN so the test is deterministic
        // regardless of the surrounding env.
        std::env::remove_var("SENTRY_DSN");
        let _guard = init(LoggerConfig {
            repo: "test-repo",
            tool: "test-tool",
            sentry_dsn: None,
            run_id: Some("test-run".into()),
        });
    }

    #[test]
    fn init_generates_run_id_when_none() {
        std::env::remove_var("SENTRY_DSN");
        // Calling without a run_id should not panic; UUID generation is the
        // only side effect we can observe externally without a custom writer.
        let _guard = init(LoggerConfig {
            repo: "test-repo",
            tool: "test-tool",
            sentry_dsn: None,
            run_id: None,
        });
    }

    #[test]
    fn init_idempotent() {
        std::env::remove_var("SENTRY_DSN");
        let _g1 = init(LoggerConfig {
            repo: "r1",
            tool: "t1",
            sentry_dsn: None,
            run_id: None,
        });
        let _g2 = init(LoggerConfig {
            repo: "r2",
            tool: "t2",
            sentry_dsn: None,
            run_id: None,
        });
    }
}
